---
title: "Methylation QC"
author: "Carlos Ruiz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Introduction

In this document, we will explore the general patterns of the methylation data.


```{r}
library(meffil)
library(minfi)
library(tidyverse)
library(dendextend)
```

## Quality control

```{r}
version <- "2021-05-21"
load(paste0("../results/preprocess/full/", version, "/chd9_dataset.normalizedBeta.GenomiRatioSet.Rdata"))
pr <- nrow(gset)
getNumLines <- function(path){
  com <- system(paste("wc -l", path), intern = TRUE)
  as.numeric(strsplit(com, " ")[[1]][1])
}
ini <- getNumLines("../data/rawdata/data.txt")
```
Our initial dataset consisted of `r as.character(ini)` probes and `r ncol(gset)` samples. Given that we only had access to detection p-values, QC was only based on this parameter. We discarded the `r ini - pr` probes with a detection p-value > 2e-16 in more than 95%, resulting in `r pr`. No samples were discarded due to low call rate.   

```{r}
load(paste0("../results/preprocess/full/", version, "/chd9_dataset.autosomic.filterAnnotatedProbes.GenomicRatioSet.Rdata"))
fin <- nrow(gset)
```

Next, we removed `r pr - fin` probes not measuring methylation, crosshibridizing or with SNPs or in sexual chromosomes. Our final dataset had `r ncol(gset)` and `r fin` probes. 

## Samples descriptives

```{r}
getSum <- function(vec, type = "continuous"){
  if (type == "continuous"){
    c(median = median(vec, na.rm = TRUE), 
      range = quantile(vec, probs = c(0.25, 0.75), na.rm = TRUE))
  } else if (type == "categorical"){
    t <- table(vec)
    data.frame(Categories = names(t),  N = as.vector(t), 
               Percentage = sprintf("%.2f", as.vector(prop.table(t)*100)))
  }
}
```

```{r}
getSum(gset$Estatus, "categorical")
```

We have 10 probands and 13 control relatives.

```{r}
getSum(gset$Sexo, "categorical")
```

We have a higher proportion of males than females.

```{r}
table(gset$Estatus, gset$Sexo)
```

Only one probands is a girl, while the controls are the mother or the father.


```{r}
table(gset$Grupo2)
```
We have two individuals with CHD9 mutations, two with REPS2 and 6 with other mutations. 5 parents were carriers of pathogenic mutations while 8 were controls.


```{r}
getSum(gset$Sentrix.Barcode, "categorical")

```

Individuals were processed in three arrays.

```{r}
table(gset$Familia, gset$Sentrix.Barcode)
```

Each family was processed in a different array.

## Global methylation patterns

We computed a PCA including the top 40000 most variable probes, in order to observe global patterns in the data.

```{r}
pcs <- meffil.methylation.pcs(getBeta(gset), probe.range = 40e3)
pcs_df <- pcs %>%
  data.frame() %>%
  mutate(SampleID = rownames(.)) %>%
  tibble() %>%
  left_join(data.frame(colData(gset)), by = "SampleID") %>%
  mutate(Sentrix.Barcode = factor(Sentrix.Barcode))


vars <- c("Sentrix.Barcode", "Sexo", "Estatus", "Familia", "Grupo", "Grupo2")
names(vars) <- vars
pcsLM <- lapply(vars, function(x) {
  lapply(paste0("PC", 1:10), function(y) summary(lm(formula(paste(y, "~ ", x)), pcs_df)))
})

pcsP <- sapply(pcsLM, function(x) {
  sapply(x, function(y) pf(y$fstatistic[1], y$fstatistic[2], y$fstatistic[3], lower.tail = FALSE))
})
rownames(pcsP) <- paste0("PC", 1:10)
pcsP
min(pcsP)
1 - max(pcsP)
```

Family has the strongest association with the general methylation patterns. Sentrix barcode is also associated with the second and fifth PCs.

```{r}
ggplot(pcs_df, aes(x = PC1, y = PC2, color = Sentrix.Barcode)) +
         geom_point() +
  theme_bw()
```

Some individuals from the same array have extreme values for PCs 1 and 2. In addition, individuals from one barcode are clustered in top left corner.

```{r}
ggplot(pcs_df, aes(x = PC2, y = PC5, color = Sentrix.Barcode)) +
         geom_point() +
  theme_bw()
```

Using PCs 2 and 5 we can easily separate individuals by array.

```{r}
ggplot(pcs_df, aes(x = PC1, y = PC2, color = Sexo)) +
         geom_point() +
  theme_bw()
```

We do not observe differences in global methylation due to sex.


```{r}
ggplot(pcs_df, aes(x = PC1, y = PC2, color = Estatus)) +
         geom_point() +
  theme_bw()
```

We did not observe differences between probands and their parents.

```{r}
ggplot(pcs_df, aes(x = PC1, y = PC2, color = Familia)) +
         geom_point() +
  theme_bw()
```

Individuals from the same family, were processed in the same batch, the first two components also separated individuals by family. 

```{r}
ggplot(pcs_df, aes(x = PC2, y = PC6, color = Familia)) +
         geom_point() +
  theme_bw()
```
With second and sixth PC we can clearly differentiate the families.

```{r}
ggplot(pcs_df, aes(x = PC1, y = PC2, color = Familia, shape = Grupo2)) +
         geom_point(size = 3) +
  theme_bw()
```

We do not observe a clear pattern differentiating cases and controls. 